### PR TITLE
Make URI Unc path recognition more consistent across platforms

### LINKED ISSUE #33049: [Uri] UNC paths should consistently convert backslashes
Related https://github.com/dotnet/runtime/issues/22103 https://github.com/dotnet/runtime/issues/33029

UNC paths passed to the Uri constructor seem to receive the same platform dependent treatment as file schemes, yet UNC is firmly a Windows file path concept. I believe backslashes should always be replaced in UNC paths, also when running on Unix.

```cs
var path = "test\\foo/bar";
var x = new Uri("//example.com/" + path);
Console.WriteLine(x.GetComponents(UriComponents.AbsoluteUri, UriFormat.Unescaped));
```

This outputs `test\foo/bar` on Unix and `test/foo/bar` on Windows.

And as alluded to in https://github.com/dotnet/runtime/issues/33029 I really think there should be a way to roundtrip file scheme urls between Unix and Windows; without having to deal with (or know about) these platform specific shenanigans.

Appendix E.4 of the file scheme RFC states https://tools.ietf.org/html/rfc8089#appendix-E.4

>It may be possible to translate or update such an invalid file URI by
replacing all backslashes "\" with slashes "/" if it can be
determined with reasonable certainty that the backslashes are
intended as path separators.

Though what should reasonable certainty mean? I would say that converting backslashes to slashes for ANY host doesn't sound like you could be reasonably certain about that transformation being correct, a gigantic portion of the internet is Unix based. 

Reasonably certain to me sounds more like, this file url is local like `file:///file.txt`, `file://localhost/file.txt` or `file://resolvestothishost/file.txt` or you have specific knowledge (a host list) that it is a windows machine, and the RFC would back me up on that local definition.

https://tools.ietf.org/html/rfc8089#section-1.1

>The term "local file URI" is used to describe file URIs that have no
"authority" component or where the authority is the special string
"localhost" or a fully qualified domain name that resolves to the
machine from which the URI is being interpreted (Section 2).

Now I understand it would be a breaking change to change `file` semantics for Uri, but I implore you to look into allowing this to be configured.

### LINKED ISSUE #33053: [Uri] Localpath shouldn't be encoding ucschars
Passing Uri a UNC path that contains unicode special characters - which are valid directory and file names on Windows - get encoded to a percent encoded triplet anyway. This is correct when dealing with Uris, but `Uri.LocalPath` should not encode these characters.

```cs
var path = "\uFFF0";
var x = new Uri("//example/" + path);
Console.WriteLine(x.LocalPath);
```

Outputs `//example/%EF%BF%B0` yet it should output `//example/￰￰￰` (which is `//example/\uFFF0`)

### PR DESCRIPTION
**Pull request by MihaZupan**
Fixes #33049, fixes #33053, contributes to #33029.

While UNC is defined as starting with two backslashes, Windows does sometimes recognize the forward-slash form as well.

This PR makes Uri recognize strings starting with two forward slashes as UNC on Unix to make behavior consistent across platforms.

Previously:

Input | Windows | Unix
-- | -- | --
\\\\foo | UNC | UNC
//foo | UNC | Absolute file path
file://\\\\foo | UNC | UNC
file:////foo | UNC | UNC

Now, these are all treated as UNC (including when the two slashes are mixed - `"/\foo"` or `"\/foo"`).

This is a breaking change as paths starting with two forward slashes are no longer recognized as absolute file paths on Unix. Being recognized as UNC also changes their behavior to replace `/` with `\` when looking at the `LocalPath` property.

### CODE REVIEW

**Comment by MihaZupan** [comment]
cc @NinoFloris

File: src/libraries/System.Private.Uri/src/System/Uri.cs [hunk 1]
```diff
@@ -3640,9 +3640,13 @@ private unsafe void ParseRemaining()
             // Unix: Unix path?
             if (!IsWindowsSystem && idx < length && uriString[idx] == '/')
             {
-                flags |= (Flags.UnixPath | Flags.ImplicitFile | Flags.AuthorityFound);
-                syntax = UriParser.UnixFileUri;
-                return idx;
+                // A path starting with 2 / or \ (including mixed) is treated as UNC and will be matched below
+                if (idx + 1 == length || (uriString[idx + 1] != '/' && uriString[idx + 1] != '\\'))
```
Comment (scalablecory) [inline src/libraries/System.Private.Uri/src/System/Uri.cs:3644]: Can this `if` be merged with the above? Seems no point to have a nested `if` here. Elsewhere in code we break multi-test `if` statements into multiple lines but one `if` statement.

**Review by scalablecory — APPROVED** [review]
lgtm, thanks.

File: src/libraries/System.Private.Uri/src/System/Uri.cs (hunk 1 above)
Comment (MihaZupan) [reply src/libraries/System.Private.Uri/src/System/Uri.cs:3644]: No point other than readability (being able to position the comment more cleanly), I'll merge them.

**Comment by danmosemsft** [comment]
@MihaZupan this is labeled breaking change. Can you please open an issue with https://github.com/dotnet/docs/issues/new?template=dotnet-breaking-change.md 

I don't [see an existing one](https://github.com/dotnet/docs/issues?q=+is%3Aissue+label%3Abreaking-change+)

**Pull request merged**